File: 4_flask/src/systems/Dice.py
import re
import random

BEAST = r'(?P<count>\d+)d(?P<dice>[\d]+)(?P<bonus>[+-]\d+)?'

def roll_dice(dice_str):
  results = re.match(BEAST, dice_str)
  count = int(results.group("count"))
  dice = int(results.group("dice"))
  bonus = int(results.group("bonus") or 0)
  rolls = [random.randint(1,dice) for i in range(count)]
  total = sum(rolls) + bonus
  return {"total": total, "rolls": rolls, "bonus": bonus}

def rand_from_range(dmg_str):
  vals = [int(x) for x in dmg_str.split('-')]
  return { "total": random.randint(*vals) }

File: week_1_13/06_golang/go/src/battle-sim/systems/Dice.go
package dice

import (
	"math/rand"
	"regexp"
	"strconv"
	"strings"
)

var regex, _ = regexp.Compile(`(?P<count>\d+)d(?P<dice>[\d]+)(?P<bonus>[+-]\d+)?`)

func RandInt(min, max int) int {
	diff := max - min
	randint := rand.Intn(diff) + min
	return randint
}

func RollDice(roll string) map[string]int {
	match := regex.FindStringSubmatch(roll)
	result := make(map[string]string)
	for i, name := range regex.SubexpNames() {
		if i != 0 && name != "" {
			result[name] = match[i]
		}
	}
	count, _ := strconv.Atoi(result["count"])
	dice, _ := strconv.Atoi(result["dice"])
	elem, ok := result["bonus"]
	bonus := 0
	if ok {
		bonus, _ = strconv.Atoi(elem)
	}
	rolls := 0
	for i := 0; i < count; i++ {
		rolls += RandInt(1, dice)
	}
	dice_result := map[string]int{
		"total": rolls + bonus,
		"rolls": rolls,
		"bonus": bonus,
	}
	return dice_result
}

func RandFromRange(dmg_str string) map[string]int {
	parts := strings.Split(dmg_str, "-")
	min, _ := strconv.Atoi(parts[0])
	max, _ := strconv.Atoi(parts[1])
	result := map[string]int{
		"total": RandInt(min, max),
	}
	return result
}
